1. HistoryManager: add metods remove(int id) and removeAll().

Just-Roman opened the pull request:
2.InMemoryHistoryManager:
- implementation Linklist in HashMap and class Node.
- change logic add(Task task), getHistory().
- add methods:  linkLast(Task task), remove(int id), removeAll(), removeNode(Node node)
- add class Node.
3. class InMemoryTaskManagerTest:
- add methods: checkDeleteIdSubtask(), checkDeleteSubtaskForEpic(), checkChangeSetter().
4. class Main: add method testForSprint_6().

Commit:
1. HistoryManager: add metods remove(int id) and removeAll().
2.InMemoryHistoryManager:
- implementation Linklist in HashMap and class Node.
- change logic add(Task task), getHistory().
- add methods:  linkLast(Task task), remove(int id), removeAll(), removeNode(Node node)
- add class Node.
3. class InMemoryTaskManagerTest:
- add methods: checkDeleteIdSubtask(), checkDeleteSubtaskForEpic(), checkChangeSetter().
4. class Main: add method testForSprint_6().

Commit:
1. HistoryManager: add metods remove(int id) and removeAll().
2.InMemoryHistoryManager:
- implementation Linklist in HashMap and class Node.
- change logic add(Task task), getHistory().
- add methods:  linkLast(Task task), remove(int id), removeAll(), removeNode(Node node)
- add class Node.
3. class InMemoryTaskManagerTest:
- add methods: checkDeleteIdSubtask(), checkDeleteSubtaskForEpic(), checkChangeSetter().
4. class Main: add method testForSprint_6().

slavrikk commented on src/ru/practicum/task_tracker/manager/InMemoryHistoryManager.java:
```diff
@@ -2,25 +2,101 @@
+            }
+            remove(task.getId());
+            linkLast(task);
+        } else if (memory.isEmpty()) {
```
Блок if else на строках 25-30 излишний, метод  `linkLast(task)` должен справляться с кейсами когда `first `может быть `null`

Just-Roman replied:
Спасибо! Принято, исправлю.

slavrikk commented on src/ru/practicum/task_tracker/manager/InMemoryHistoryManager.java:
```diff
@@ -2,25 +2,101 @@
+        List <Task> memoryTasks = new ArrayList<>();
+
+        if (memory.isEmpty()) {
+            return null;
```
Лучше пустой массив вернуть

Just-Roman replied:
Спасибо! Принято, исправлю.

slavrikk commented on src/ru/practicum/task_tracker/manager/InMemoryHistoryManager.java:
```diff
@@ -2,25 +2,101 @@
+        if (memory.isEmpty()) {
+            return null;
+        } else {
+            Node node = last;
```
Лучше с начала начинать

Just-Roman replied:
Тогда вывод истории получится наоборот (самая старая просматриваемая будет в списке истории первой)

slavrikk replied:
В этом и есть смысл) самая старая история должна идти первой

Just-Roman replied:
Принято! Исправил))

slavrikk commented on test/ru/practicum/task_tracker/manager/InMemoryTaskManagerTest.java:
```diff
@@ -115,4 +118,59 @@ void checkSaveLastVersionHistoryManager() {
+    @Test
+    void checkChangeSetter() {
+         /*С помощью сеттеров экземпляры задач позволяют изменить любое своё поле,
+         но это может повлиять на данные внутри менеджера.
```
Как вариант использовать final поля в тасках, при создании все будет ок, при обновлении придется создавать новую сущность.
Можешь оставить как есть, не принципиально

Just-Roman replied:
Спасибо! Принято, пока оставлю так.

Commit:
1. InMemoryHistoryManager:
- method linkLast(Task task) work in if (first == null).
- method getHistory() all return memoryTasks.

Commit:
1. InMemoryHistoryManager:
- method linkLast(Task task) work in if (first == null).
- method getHistory() all return memoryTasks.

Just-Roman commented:
Вообще не понимаю на что ругается Git, буду благодарен если подскажете.

slavrikk commented on src/ru/practicum/task_tracker/manager/InMemoryHistoryManager.java:
```diff
@@ -2,25 +2,103 @@
     public void add(Task task) {
-        if (memory.size() == 10) {
-            memory.removeFirst();
+        if (memory.containsKey(task.getId())) {
```
linkLast можно вынести из if/else блока:

   ```
     if (memory.containsKey(task.getId())) {
            if (memory.size() == 1) {
                return;
            } else if (memory.get(task.getId()) == last) {
                last = last.prev;
            } else if (memory.get(task.getId()) == first) {
                first = first.next;
            }
            remove(task.getId());
        }  
         linkLast(task);

```

Just-Roman replied:
Спасибо! Не доглядел, исправлю.

slavrikk reviewed (commented):
Следует поправить реформатирование:
![image](https://github.com/Just-Roman/java-kanban/assets/58359431/033aa468-5306-411e-b529-44844d1af17c)

Commit:
1. InMemoryHistoryManager:
- removed the method from the block else if.

Commit:
1. InMemoryHistoryManager:
- removed the method from the block else if.

Commit:
added a space

Commit:
added a space class Main

Commit:
added a space class InMemoryHistoryManager

Commit:
added a space interface HistoryManager
